File: 백준/Gold/5550. 헌책방/헌책방.java
/**
 * Author: nowalex322, Kim HyeonJae
 */
import java.io.*;
import java.util.*;

public class Main {
    static BufferedReader br;
    static BufferedWriter bw;
    static StringTokenizer st;
    static int N, K, bookCnt, categorySize;
    static long dp[][];
    static ArrayList<Integer>[] books;

    public static void main(String[] args) throws Exception {
        new Main().solution();
    }

    public void solution() throws Exception {
        br = new BufferedReader(new InputStreamReader(System.in));
//        br = new BufferedReader(new InputStreamReader(new FileInputStream("src/main/java/BOJ_5550_헌책방/input.txt")));
        bw = new BufferedWriter(new OutputStreamWriter(System.out));

        st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        K = Integer.parseInt(st.nextToken());

        dp = new long[11][K + 1];

        books = new ArrayList[11];
        for (int i = 0; i < 11; i++) {
            books[i] = new ArrayList<>();
        }

        int price, category;
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            price = Integer.parseInt(st.nextToken());
            category = Integer.parseInt(st.nextToken());
            books[category].add(price);
        }

        for (int i = 1; i <= 10; i++) {
            Collections.sort(books[i], Collections.reverseOrder());

            ArrayList<Integer> tmp = new ArrayList<>(books[i]);

            for (int j = 1; j < books[i].size(); j++) {
                books[i].set(j, books[i].get(j) + books[i].get(j - 1));
            }

            for (int j = 1; j < books[i].size(); j++) {
                books[i].set(j, books[i].get(j) + j * (j + 1));
            }
        }
        
        int res = 0;
        bookCnt = 0;
        for (int i = 1; i <= 10; i++) {
            categorySize = books[i].size();
            /* 헷갈린부분
             * 똑같은 개수로 진행되어야하므로 이전에 골랐든 안골랐든 cnt만큼 추가
             * 예를 들면 앞 종류에서 2개 뒷종류에서 3개랑 같은 선상에서 뒷종류만 5개를 고른경우가 비교되어야하므로 min(K, size)하면안됨! 놓치는 경우생김
             */
            int leftAmount = Math.min(K, bookCnt + categorySize);
            for (int j = 0; j <= leftAmount; j++) {
                dp[i][j] = dp[i - 1][j];
                for (int k = 1; k <= categorySize; k++) {
                    if (j >= k) dp[i][j] = Math.max(dp[i][j], dp[i - 1][j - k] + books[i].get(k - 1));
                }
            }
            bookCnt = Math.min(bookCnt + categorySize, K);
        }

        bw.write(String.valueOf(dp[10][K]));
        bw.flush();
        bw.close();
        br.close();
    }
}

File: 백준/Gold/5550. 헌책방/헌책방.cc
/**
 * Author: nowalex322, Kim HyeonJae
 */
#include <bits/stdc++.h>
using namespace std;

// #define int long long
#define MOD 1000000007
#define INF LLONG_MAX
#define ALL(v) v.begin(), v.end()

#ifdef LOCAL
#include "algo/debug.h"
#else
#define debug(...) 42
#endif

void solve() {
    int N, K;
    cin >> N >> K;
    vector<vector<int>> books(11);
    vector<vector<long long>> dp(11, vector<long long>(K + 1));

    for (int i = 0; i < N; i++) {
        int price, category;
        cin >> price >> category;
        books[category].push_back(price);
    }

    for (int i = 1; i <= 10; i++) {
        sort(books[i].rbegin(), books[i].rend());
        // sort(books[i].begin(), books[i].end(), greater<int>());
        // sort(books[i].begin(), books[i].end(),
        //      [](int a, int b) { return a > b; });

        vector<int> tmp = books[i];
        for (int j = 1; j < books[i].size(); j++) {
            books[i][j] = books[i][j] + books[i][j - 1];
        }
        for (int j = 1; j < books[i].size(); j++) {
            books[i][j] = books[i][j] + j * (j + 1);
        }
    }

    int bookCnt = 0;
    for (int i = 1; i <= 10; i++) {
        int categorySize = books[i].size();
        int leftAmount =
            min(K, bookCnt + categorySize);  // 고를 수 있는 최대 수

        for (int j = 0; j <= leftAmount; j++) {
            dp[i][j] = dp[i - 1][j];
            for (int k = 1; k <= categorySize; k++) {
                if (j >= k) {
                    dp[i][j] =
                        max(dp[i][j], dp[i - 1][j - k] + books[i][k - 1]);
                }
            }
        }
        bookCnt = min(bookCnt + categorySize, K);
    }
    cout << dp[10][K] << "\n";
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int tt = 1;  // 기본적으로 1번의 테스트 케이스를 처리
    // cin >> tt;    // 테스트 케이스 수 입력 (필요 시)

    while (tt--) {
        solve();
    }
    return 0;
}